### ISSUE TITLE
Mac downloads the update, but does not install it

### ISSUE THREAD

**Issue by alfredsgenkins** [electron-userland/electron-builder#6058]
* **Electron-Builder Version**: `^22.10.5`

* **Node Version**: `v14.17.0`

* **Electron Version**: `^11.0.1`
* **Electron Type (current, beta, nightly)**: `current`

* **Target**: `dmg`

Hey!

This issue is similar by description to https://github.com/electron-userland/electron-builder/issues/2317, but I have not found any answers there. I decided to open new issue and provide more background.

My application on Mac shows the following notification:

<img width="336" alt="Screenshot 2021-07-14 at 16 23 34" src="https://user-images.githubusercontent.com/29531824/125629621-05283606-6699-40a6-a257-6401c3365f53.png">

But once I close and reopen the app back nothing happens. The old version is still displayed. I also want my application use the `quitAndInstall` pattern. For me the fact the the notification is displayed in the first place is strange.

**The background**:
- I have configured automatic updates for my application
- I have made my application notarized by apple
- I want to use `quitAndInstall`, I have ensure the safe exit
- I am using Private GitHub repository to store releases
- Autoupdate on Linux AppImage works

**How my application handles an update**:

```typescript
import { app, BrowserWindow, shell } from 'electron';
import { autoUpdater } from 'electron-updater';
import log from 'electron-log';

function ensureSafeQuitAndInstall() {
  app.removeAllListeners('window-all-closed');
  const browserWindows = BrowserWindow.getAllWindows();
  browserWindows.forEach((browserWindow) => {
    browserWindow.removeAllListeners('close');
  });
  autoUpdater.quitAndInstall(false);
}

export default class AppUpdater {
  constructor() {
    log.transports.file.level = 'info';
    autoUpdater.logger = log;
    autoUpdater.checkForUpdatesAndNotify();

    autoUpdater.on('update-downloaded', () => {
      setImmediate(ensureSafeQuitAndInstall);
    });
  }
}
```

I took the code sample from this issue and built my implementation on top if it: https://github.com/electron-userland/electron-builder/issues/1604

**Why I think the signature is fine**:

<img width="471" alt="Screenshot 2021-07-14 at 16 32 21" src="https://user-images.githubusercontent.com/29531824/125630901-084567e5-dd4a-4f28-b29d-3951e7840b73.png">

Despite me expecting the signature with the Developer ID, the app only shows notarization and hardening. Not sure why. The app however, only promts the this on first install, thus I assume it is OK:

<img width="422" alt="Screenshot 2021-07-14 at 16 36 28" src="https://user-images.githubusercontent.com/29531824/125631494-2f2d9842-1543-4d20-a101-a245bd47b4c7.png">

**How I sign my app**:

I used following configuration in `package.json`:

```json
  ...
  "afterSign": ".erb/scripts/Notarize.js",
  "mac": {
    ...
    "target": [
      "dmg"
    ],
    "asarUnpack": "**/*.node",
    "type": "distribution",
    "hardenedRuntime" : true,
    "gatekeeperAssess": false,
    "entitlements": "assets/entitlements.mac.plist",
    "entitlementsInherit": "assets/entitlements.mac.plist",
    "icon": "assets/icons/mac/icon.icns",
    "provisioningProfile": "assets/preflight_provision_profile.provisionprofile"
  },
  ...
```

My notarization script is using `electron-notarize` and I sign it with following code:

```typescript
const { notarize } = require('electron-notarize');
const { build } = require('../../package.json');

exports.default = async function notarizeMacos(context) {
  const { electronPlatformName, appOutDir } = context;
  const appName = context.packager.appInfo.productFilename;
  
  await notarize({
    appBundleId: build.appId,
    appPath: `${appOutDir}/${appName}.app`,
    appleId: process.env.APPLE_ID,
    appleIdPassword: process.env.APPLE_ID_PASS,
    ascProvider: process.env.APPLE_PROVIDER_ID,
  });
};
```

I build my app using following command:

```bash
CSC_LINK="XXX" \
CSC_KEY_PASSWORD=XXX \
APPLE_ID="XXX" \
APPLE_ID_PASS="XXX" \
APPLE_PROVIDER_ID="XXX" \
yarn electron-builder build -mwl --publish always
```

**What I also suspect**:

The react boilerplate I use, contained the following code:

```typescript
app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});
```

I heared that this could prevent app from closing. But I call `app.removeAllListeners('window-all-closed')` thus this should not be the case.

In logs, I see following message (relating me to https://github.com/electron-userland/electron-builder/issues/486, but again, this appears to not break anything):

```bash
[XXX] [warn] Error: ditto: Couldn't read PKZip signature
[XXX] [error] Error: Error: ditto: Couldn't read PKZip signature
```

Any ideas, suggestions?

**Comment by mmaietta** [comment]
That may have been resolved through https://github.com/electron-userland/electron-builder/pull/5802
Please try latest electron-builder at 22.11.9

**Comment by cuantmac** [comment]
> That may have been resolved through #5802
> Please try latest electron-builder at 22.11.9

nothing happend

**Comment by vraravam** [comment]
I'm on v22.12.1 - same case (nothing fixed)

**Comment by jmeinke** [comment]
We can confirm this issue. It might have something to do with missing administration privileges for being able to update the application. After entering user credentials for allowing the update of the app, then executing the update again it works.

**Comment by nicograef** [comment]
@Nokel81 Is this a connected (or the same) bug that you fixed with #6390 ?
Thanks!

**Comment by Nokel81** [comment]
@nicograef It might be. I have not had this problem since using that fix I provided.

**Comment by theolivenbaum** [comment]
We're seeing the same issue on our app. Update is downloaded, but never installed. Used to work before until recently.

**Comment by Nokel81** [comment]
@theolivenbaum Are you using `electron-updater@4.6.2`?

**Comment by theolivenbaum** [comment]
Seems like 4.6.1 https://github.com/theolivenbaum/Electron.NET/blob/21427f0b449ba8e4d37e02b037685bb5e9a4d323/ElectronNET.Host/package-lock.json#L411

Was this fixed on 4.6.2? If so I'll give it a try tomorrow!

**Comment by Nokel81** [comment]
Yes, my PR was included in 4.6.2 according to the comment posted onto it

**Comment by helious** [comment]
@alfredsgenkins I think I've figured this out.. after stumbling around the myriad threads on this issue, it's a simple configuration setting 😆 (https://www.electron.build/configuration/mac#MacConfiguration-target)

so just doing this:
```json
"mac": {
    ...
    "target": [
      "default"
    ],
```

as the the auto updater relies on the `.zip`, just having `"dmg"` there won't sign the zip; but `"default" signs the `.dmg` and `.zip`.

the logs were able to point me in the right direction:
```
[2022-02-27 13:55:15.447] [info] Found version 0.0.1 (url: app-0.0.1.dmg)
[2022-02-27 13:55:15.448] [info] Downloading update from app-0.0.1.dmg
[2022-02-27 13:55:15.637] [info] Update has already been downloaded to /Users/user/Library/Application Support/Caches/app-updater/pending/update.zip).
[2022-02-27 13:55:15.657] [info] / requested
[2022-02-27 13:55:15.661] [info] /1645988115637-3819.zip requested
[2022-02-27 13:55:15.661] [info] /1645988115637-3819.zip requested by Squirrel.Mac, pipe /Users/user/Library/Application Support/Caches/app-updater/pending/update.zip

**Comment by vendethiel** [comment]
Relevant

**Comment by liangskyli** [comment]
I also have the same problem

**Comment by willemdjong** [comment]
I ran into the same issue: the app successfully downloads a new version, but after calling quitAndRestart(), it quits and restarts with the same version, not the new one. Only after calling autoUpdater.quitAndRestart() a second time does the new version actually get installed.
